csharp/extractor/Semmle.Extraction.CSharp/Entities/Types/Type.cs

```diff
@@ -347,6 +355,211 @@ public override bool Equals(object? obj)
+                var workQueue = new Queue<INamedTypeSymbol>();
+                workQueue.Enqueue(startSymbol);
+
+                while (workQueue.Count > 0)
+                {
+                    var current = workQueue.Dequeue();
+                    if (instantiationClosure.Contains(current) ||
+                        !current.IsGenericType)
+                    {
+                        continue;
+                    }
+
+                    instantiationClosure.Add(current);
+
+                    if (SymbolEqualityComparer.Default.Equals(current, current.OriginalDefinition))
+                    {
+                        // Definition, so enqueue all base types and interfaces
+                        if (current.BaseType != null)
+                        {
+                            workQueue.Enqueue(current.BaseType);
+                        }
+
+                        foreach (var i in current.Interfaces)
+                        {
+                            workQueue.Enqueue(i);
+                        }
+                    }
+                    else
+                    {
+                        // Reference, so enqueue all type arguments and their original definitions:
+                        foreach (var namedTypeArgument in current.TypeArguments.OfType<INamedTypeSymbol>())
+                        {
+                            workQueue.Enqueue(namedTypeArgument);
+                            workQueue.Enqueue(namedTypeArgument.OriginalDefinition);
+                        }
+                    }
+                }
+            }
+
+            private bool HasExpandingCycle()
+            {
+                return startSymbol.TypeParameters.Any(HasExpandingCycle);
+            }
+
+            private bool HasExpandingCycle(ITypeParameterSymbol start)
+            {
+                var visited = new HashSet<ITypeParameterSymbol>(SymbolEqualityComparer.Default);
+                var recStack = new HashSet<ITypeParameterSymbol>(SymbolEqualityComparer.Default);
+                var hasExpandingCycle = HasExpandingCycle(start, visited, recStack, start, hasSeenExpandingEdge: false);
+                return hasExpandingCycle;
+            }
+
+            private List<(ITypeParameterSymbol To, bool IsExpanding)> GetOutgoingEdges(ITypeParameterSymbol typeParameter)
+            {
+                return edges.TryGetValue(typeParameter, out var outgoingEdges)
+                    ? outgoingEdges
+                    : new List<(ITypeParameterSymbol, bool)>();
+            }
+
+            /// <summary>
+            /// A modified cycle detection algorithm based on DFS.
+            /// </summary>
+            /// <param name="current">The current node that is being visited</param>
+            /// <param name="visited">The nodes that have already been visited by any path.</param>
+            /// <param name="currentPath">The nodes already visited on the current path. Could be a List<> if the order was important.</param>
```

Comment:
Is it actually enough to record just the size of the current path, and not the actual elements on the path?

Reply:
We do record the elements in the path, but not their order. `currentPath` is a `HashSet<ITypeParameterSymbol>`.

Reply:
Ahh, I understand what you mean. I think it's easier to understand that this is a slightly modified cycle detection algorithm, if we keep the `currentPath`. I'm making it a `List<>` and removing the `start` parameter, as we know that the start is going to be the first item in the list. What do you think?